File: cpp/pool/pool.cpp
#include "pool.hpp"

#include "../../c/utils/args_parser.h"

#include <algorithm>

Pool::Pool(Int capacity) {
    self.blocks = std::deque<Block>();
    self.capacity = capacity;
}

Block &Pool::add_block(usize size, SInt ttl) {
    if (self.count() >= self.capacity) {
        panic("Pool is at capacity. Cannot allocate new blocks.");
    }

    return self.blocks.emplace_back(size, ttl);
}

void Pool::del_block(Policy policy, Random &rng) {
    if (policy == POLICY_NEVER) {
        return;
    }

    if (self.blocks.size() == 0) {
        return;
    }

    switch (policy) {
    case POLICY_LIFO:
        self.blocks.pop_back();
        break;
    case POLICY_FIFO:
        self.blocks.pop_front();
        break;
    case POLICY_RANDOM: {
        Int idx = rng.uniform(0, self.count());
        auto p = self.blocks.begin();
        for (Int i = 0; i < idx; i++, p++)
            ;
        self.blocks.erase(p);
    } break;
    case POLICY_BIG_FIRST: {
        auto it = std::max_element(
            self.blocks.begin(), self.blocks.end(),
            [](Block &lhs, Block &rhs) { return lhs.size < rhs.size; });

        self.blocks.erase(it);
    } break;
    case POLICY_SMALL_FIRST: {
        auto it = std::min_element(
            self.blocks.begin(), self.blocks.end(),
            [](Block &lhs, Block &rhs) { return lhs.size < rhs.size; });

        self.blocks.erase(it);
    } break;
    default:
        panic("Unknown policy");
    }
}

void Pool::update_and_prune() {
    auto it = self.blocks.begin();
    while (it != self.blocks.end()) {
        if (it->ttl > 0) {
            it->ttl--;
        }

        if (it->ttl == 0) {
            log_info(ANSI_COLOR(RED, "Pruning") "    " DBG_BLOCK_STR,
                     size_str(it->size, 2), it->ttl_org);
            it = self.blocks.erase(it);
        } else {
            it++;
        }
    }
}

void Pool::touch_pages(Block &block, Int stride) const {
    Int rnd_val = 0xA5;
    u8 *data = block.data.data();
    for (Int off = 0; off < block.size; off += stride) {
        ((volatile u8 *)data)[off] = static_cast<u8>(rnd_val);
        rnd_val = rnd_val * 0x5DEECE66D + 0xB;
    }
}

File: rust/src/pool.rs
use std::collections::VecDeque;

use crate::random::Random;
use crate::tracker::GLOBAL_TRACKER;
use crate::utils::args_parser::Policy;

pub struct Block {
    data: Vec<u8>,
    ttl: isize,
    ttl_org: isize,
}

impl Drop for Block {
    fn drop(&mut self) {
        unsafe {
            GLOBAL_TRACKER.remove_alloc(self.data.capacity());
        }
    }
}

impl Block {
    pub fn new(size: usize) -> Self {
        Self::with_ttl(size, -1)
    }

    pub fn with_ttl(size: usize, ttl: isize) -> Self {
        unsafe {
            GLOBAL_TRACKER.add_alloc(size);
        }
        Self {
            data: Vec::with_capacity(size),
            ttl,
            ttl_org: ttl,
        }
    }

    #[inline(always)]
    pub fn tick(&mut self) {
        if self.ttl > 0 {
            self.ttl -= 1;
        }
    }

    #[inline(always)]
    pub const fn alive(&self) -> bool {
        self.ttl < 0 || self.ttl > 0
    }

    #[inline(always)]
    pub const fn ttl_org(&self) -> isize {
        self.ttl_org
    }

    pub fn touch(mut self, stride: usize) -> Self {
        let ptr = self.data.as_mut_ptr();

        let mut i = 0usize;
        let mut rnd_val: usize = 0x5A;
        while i < self.data.capacity() {
            unsafe {
                // Safe because we are only touching Vec's allocated space
                *ptr.add(i) = rnd_val as u8;
            }
            rnd_val = rnd_val.wrapping_mul(0x5DEECE66D).wrapping_add(0xB);
            i += stride;
        }

        self
    }
}

pub struct Pool {
    blocks: VecDeque<Block>,
    pub capacity: usize,
}

impl Pool {
    pub fn new(capacity: usize) -> Self {
        Self {
            blocks: VecDeque::with_capacity(capacity),
            capacity: capacity,
        }
    }

    pub fn add_block(&mut self, size: usize, stride: usize) {
        if self.blocks.len() >= self.capacity {
            return;
        }
        self.blocks.push_back(Block::new(size).touch(stride));
    }

    pub fn add_block_with_ttl(&mut self, size: usize, ttl: isize, stride: usize) {
        if self.blocks.len() >= self.capacity {
            return;
        }
        self.blocks
            .push_back(Block::with_ttl(size, ttl).touch(stride));
    }

    pub fn del_block(&mut self, policy: Policy, rng: &mut Random) {
        match policy {
            Policy::Lifo => self.blocks.pop_back(),
            Policy::Fifo => self.blocks.pop_front(),
            Policy::Random => self.blocks.remove(rng.uniform(0, self.blocks.len())),
            Policy::BigFirst => self.blocks.remove(
                self.blocks
                    .iter()
                    .enumerate()
                    .max_by_key(|(_, b)| b.data.capacity())
                    .map(|(i, _)| i)
                    .unwrap_or(0),
            ),
            Policy::SmallFirst => self.blocks.remove(
                self.blocks
                    .iter()
                    .enumerate()
                    .min_by_key(|(_, b)| b.data.capacity())
                    .map(|(i, _)| i)
                    .unwrap_or(0),
            ),
            Policy::Never => None
        };
    }

    pub fn update_and_prune(&mut self) {
        let mut i = 0usize;
        while i < self.blocks.len() {
            self.blocks[i].tick();

            if !self.blocks[i].alive() {
                self.blocks.remove(i);
            } else {
                i += 1;
            }
        }
    }

    pub fn count(&self) -> usize {
        self.blocks.len()
    }
}
